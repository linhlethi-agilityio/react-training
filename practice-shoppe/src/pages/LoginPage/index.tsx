import { Layout } from '@layouts/';
import StyledText from '@components/Text';
import { StyledInputControl } from '@components/InputControl';
import { StyledButton } from '@components/Button';
import styled from 'styled-components';
import { FormEvent, useCallback, useRef, useState } from 'react';
import { Navigate } from 'react-router-dom';
import { StyledLoadingSpinner } from '@components/LoadingSpinner';
import { useAuth } from '@hooks/useAuth';

const LoginPage = () => {
  const [loading, setLoading] = useState<boolean>(false);

  const passwordRef = useRef<HTMLInputElement>();
  const emailRef = useRef<HTMLInputElement>();

  const { data: user, error, login } = useAuth();

  const handleLogin = useCallback(
    async (e: FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      const userInput = {
        email: emailRef.current!.value,
        password: passwordRef.current!.value,
      };

      login(userInput);
    },
    [setLoading, emailRef, passwordRef],
  );

  if (user) {
    return <Navigate to="/" replace={false} />;
  }

  return (
    <Layout>
      <StyledInputForm>
        <form onSubmit={handleLogin}>
          <StyledText as="h2" fontSize={33} lineHeight={43} fontWeight={500} textAlign="center">
            Login
          </StyledText>
          <StyledInputControl
            label="Email"
            name="email"
            id="email"
            inputType="email"
            disabled={loading}
            ref={emailRef}
          />
          <StyledInputControl
            label="Password"
            name="password"
            id="password"
            inputType="password"
            disabled={loading}
            ref={passwordRef}
          />
          {error?.body && (
            <StyledText textAlign="left" mTop={7} color="error">
              {error?.body}
            </StyledText>
          )}
          {loading && <StyledLoadingSpinner />}
          <StyledButton variant="primary" type="submit" disabled={loading}>
            Submit
          </StyledButton>
        </form>
      </StyledInputForm>
    </Layout>
  );
};

const StyledInputForm = styled.section`
  padding-top: 200px;
  padding-bottom: 252px;

  form {
    text-align: center;
    margin-top: 100px;
    width: 400px;
    border: 2px solid #d8d8d8;
    border-radius: 8px;
    padding: 30px 40px;
    margin: 100px auto 250px auto;
  }

  button {
    margin-top: 40px;
  }

  div:nth-child(n + 2) {
    margin-top: 15px;
  }
`;

export default LoginPage;
