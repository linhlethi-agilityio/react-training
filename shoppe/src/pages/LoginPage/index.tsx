import { Layout } from '@layouts/';
import StyledText from '@components/Text';
import { StyledInputControl } from '@components/InputControl';
import { StyledButton } from '@components/Button';
import styled from 'styled-components';
import { UserContext } from '@contexts/UserContext';
import { ChangeEvent, FormEvent, useCallback, useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { StyledLoadingSpinner } from '@components/LoadingSpinner';
import { login } from 'src/services/Auth';

const LoginPage = () => {
  const { state, actions } = useContext(UserContext);
  const navigate = useNavigate();

  const { userInput, errorMessage, loading } = state;

  const { setLoading, onChangeInput, setErrorMessage, postLoginRequestSuccess } = actions;

  const handleChangeInput = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;

    onChangeInput({
      [name]: value,
    });
  }, []);

  const handleLogin = useCallback(
    async (e: FormEvent<HTMLFormElement>) => {
      try {
        e.preventDefault();
        setLoading(true);
        const result = await login(userInput);

        if (result.status === 400) {
          setErrorMessage('Email or Password invalid');
          return;
        }

        const resultBody = result.body;

        postLoginRequestSuccess(resultBody.accessToken, resultBody.user);
        localStorage.setItem('login', resultBody.user.id.toString());
        navigate('/');
      } catch (error) {
        setErrorMessage(`Error! An error occurred: ${error}. Please try again later`);
      }
    },
    [userInput],
  );

  return (
    <Layout>
      <StyledInputForm>
        <form onSubmit={handleLogin}>
          <StyledText as="h2" fontSize={33} lineHeight={43} fontWeight={500} textAlign="center">
            Login
          </StyledText>
          <StyledInputControl
            label="Email"
            value={userInput.email}
            name="email"
            id="email"
            inputType="email"
            onChange={handleChangeInput}
            disabled={loading}
          />
          <StyledInputControl
            label="Password"
            value={userInput.password}
            name="password"
            id="password"
            inputType="password"
            onChange={handleChangeInput}
            disabled={loading}
          />
          {errorMessage && (
            <StyledText textAlign="left" mTop={7} color="error">
              {errorMessage}
            </StyledText>
          )}
          {loading && <StyledLoadingSpinner />}
          <StyledButton variant="primary" type="submit" disabled={loading}>
            Submit
          </StyledButton>
        </form>
      </StyledInputForm>
    </Layout>
  );
};

const StyledInputForm = styled.section`
  padding-top: 200px;
  padding-bottom: 252px;

  form {
    text-align: center;
    margin-top: 100px;
    width: 400px;
    border: 2px solid #d8d8d8;
    border-radius: 8px;
    padding: 30px 40px;
    margin: 100px auto 250px auto;
  }

  button {
    margin-top: 40px;
  }

  div:nth-child(n + 2) {
    margin-top: 15px;
  }
`;
export default LoginPage;
