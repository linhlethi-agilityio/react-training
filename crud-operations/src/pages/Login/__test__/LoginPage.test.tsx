import { render, fireEvent, act, waitFor } from '@testing-library/react';
import { ChakraProvider, theme } from '@chakra-ui/react';
import * as router from 'react-router-dom';

// Hooks
import * as hooks from '@hooks';

import LoginPage from '..';

// Mocking dependencies
jest.mock('@hooks', () => ({
  useToastCustom: jest.fn(),
  useAuth: () => ({
    getCurrentUser: jest.fn(() => ({ email: 'admin@gmail.com' })),
    loginWithEmailPassword: jest.fn(),
  }),
}));

jest.mock('jose', () => ({}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
  useLocation: jest.fn(),
}));

const customTheme = {
  ...theme,
  colors: {
    primary: '#FEAF00', // Custom colors here
  },
};

jest.mock('@constants', () => ({
  ENVS: {
    VITE_API_ENDPOINT: 'import.meta.env.VITE_API_ENDPOINT',
    VITE_SECRET_KEY: 'import.meta.env.VITE_SECRET_KEY',
  },
  ERROR_MESSAGES: {
    FIELD_REQUIRED: 'This field is required',
    MIN_PASSWORD_LENGTH: jest.fn(),
  },
  REGEX_PATTERN: {
    EMAIL:
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
  },
  ROUTERS: {
    DASHBOARD: '/',
  },
  BRAND_NAME: 'crud operations',
}));

const mockLoginWithEmailPassword = jest.fn();
const mockLogout = jest.fn();
const mockGetCurrentUser = jest.fn();
const navigate = jest.fn();
jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate);

describe('LoginPage Component', () => {
  test('renders correctly', () => {
    const { container, getByTestId } = render(
      <ChakraProvider theme={customTheme}>
        <LoginPage />
      </ChakraProvider>,
    );

    const linkElement = getByTestId('brand-logo');

    fireEvent.click(linkElement);

    expect(linkElement).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('calls login function with correct credentials on submit', async () => {
    jest.spyOn(hooks, 'useAuth').mockReturnValue({
      me: null,
      errorMessage: '',
      loginWithEmailPassword: mockLoginWithEmailPassword,
      logout: mockLogout,
      getCurrentUser: mockGetCurrentUser,
    });

    const mockReturnValue = {
      email: 'admin@gmail.com',
      password: 'Admin@1234',
    };

    const { getByLabelText, getByRole } = render(
      <ChakraProvider theme={customTheme}>
        <LoginPage />
      </ChakraProvider>,
    );

    // Click the sign in button
    act(async () => {
      // Fill in email and password fields
      fireEvent.change(getByLabelText('Email'), { target: { value: 'admin@gmail.com' } });
      fireEvent.change(getByLabelText('Password'), { target: { value: 'Admin@123' } });

      mockLoginWithEmailPassword.mockResolvedValue(mockReturnValue);

      const button = getByRole('button', { name: 'sign-in' });

      fireEvent.click(button);
    });

    waitFor(() => {
      expect(mockLoginWithEmailPassword).toHaveBeenCalled();
    });
  });
});
