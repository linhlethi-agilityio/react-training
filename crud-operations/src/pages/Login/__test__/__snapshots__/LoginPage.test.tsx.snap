// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Component renders correctly 1`] = `
<div>
  <div
    class="css-1xypjzu"
  >
    <div
      class="css-l84sf8"
    >
      <div
        class="chakra-card css-2svjbt"
      >
        <button
          class="chakra-link css-pqilmp"
          data-testid="brand-logo"
        >
          <h2
            class="chakra-heading css-1glaey6"
          >
            crud operations
          </h2>
        </button>
        <h2
          class="chakra-heading css-1omxs99"
        >
          Sign In
        </h2>
        <p
          class="chakra-text css-uxfuux"
        >
          Enter your credentials to access your account
        </p>
        <div
          class="chakra-form-control css-1h5wztp"
          role="group"
        >
          <div
            class="css-d2znx6"
          >
            <label
              class="chakra-form__label css-15k0wdv"
              for="field-:r0:"
              id="field-:r0:-label"
            >
              Email
            </label>
            <input
              class="chakra-input css-132xrut"
              id="field-:r0:"
              value=""
            />
          </div>
          <div
            class="css-d2znx6"
          >
            <label
              class="chakra-form__label css-15k0wdv"
              for="field-:r0:"
              id="field-:r0:-label"
            >
              Password
            </label>
            <input
              class="chakra-input css-132xrut"
              id="field-:r0:"
              type="password"
              value=""
            />
          </div>
          <button
            aria-label="sign-in"
            class="chakra-button css-1ib6x77"
            disabled=""
            type="button"
          >
            Sign in
          </button>
        </div>
        <p
          class="chakra-text css-1178ru4"
        >
          Forgot your password?Â 
          <button
            class="chakra-link css-bo1wqf"
          >
            <span
              class="chakra-text css-0"
            >
              Reset Password
            </span>
          </button>
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
